src/UmbracoExamine.PDF/PdfIndexPopulator.cs

```diff
@@ -32,12 +33,99 @@ public PdfIndexPopulator(IMediaService mediaService, IPdfIndexValueSetBuilder me
             RegisterIndex("PDFIndex");
+
+            //Tie into various events one media saved/trashed/deleted/restored
+            Umbraco.Core.Services.Implement.MediaService.Saved += MediaService_Saved;
+            Umbraco.Core.Services.Implement.MediaService.Deleted += MediaService_Deleted;
+            Umbraco.Core.Services.Implement.MediaService.Trashed += MediaService_Trashed;
+            Umbraco.Core.Services.Implement.MediaService.Moved += MediaService_Moved;
+        }
+
+
+        /// <summary>
+        /// If pdfs were restored form recycle bin, re-index them
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="e"></param>
+        private void MediaService_Moved(IMediaService sender, Umbraco.Core.Events.MoveEventArgs<IMedia> e)
+        {
+            var restored = e.MoveInfoCollection
+                .Where(m => m.OriginalPath.StartsWith("-1")) //recyclebin
+                .Select(m => m.Entity);
+            if (restored.Any())
+            {
+                AddToIndex(restored);
+            }
+        }
+
+        /// <summary>
+        /// Remove any pdfs that have been trashed
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="e"></param>
+        private void MediaService_Trashed(IMediaService sender, Umbraco.Core.Events.MoveEventArgs<IMedia> e)
+        {
+            RemoveFromIndex(e.MoveInfoCollection.Select(m => m.Entity));
+        }
+
+        /// <summary>
+        /// Remvoe any pdfs that have been deleted
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="e"></param>
+        private void MediaService_Deleted(IMediaService sender, Umbraco.Core.Events.DeleteEventArgs<IMedia> e)
+        {
+            RemoveFromIndex(e.DeletedEntities);
+        }
+
+        /// <summary>
+        /// if any of the media is a pdf remove it from the PDFIndex
+        /// </summary>
+        /// <param name="media"></param>
+        private void RemoveFromIndex(IEnumerable<IMedia> media)
+        {
+            if (_examineManager.TryGetIndex("PDFIndex", out var index))
+            {
+                var ids = media.Where(m => m.GetValue<string>("umbracoExtension") == "pdf").Select(m => m.Id.ToString());
+                if (ids.Any())
+                {
+                    index.DeleteFromIndex(ids);
+                }
+            }
+
+        }
+
+        /// <summary>
+        /// Index any saved pdfs
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="e"></param>
+        private void MediaService_Saved(IMediaService sender, Umbraco.Core.Events.SaveEventArgs<Umbraco.Core.Models.IMedia> e)
+        {
+            AddToIndex(e.SavedEntities);
+        }
+
+        /// <summary>
+        /// Add any media that is a pdf to the PDFIndex
+        /// </summary>
+        /// <param name="media"></param>
+        private void AddToIndex(IEnumerable<IMedia> media)
```

Comment:
I think this should be `public` and used from the component..

Reply:
```suggestion
        public void AddToIndex(IEnumerable<IMedia> media)
```

Reply:
The method is almost identical to `RemoveFromIndex`
We could handle the common stuff in a single method, that takes a delegate that removes or add's to the index